code/modules/client/asset_cache.dm

```diff
@@ -161,29 +161,186 @@ GLOBAL_LIST_EMPTY(asset_datums)
 		return new type()
 	return GLOB.asset_datums[type]
 
+/datum/asset
+	var/_abstract = /datum/asset
+
 /datum/asset/New()
 	GLOB.asset_datums[type] = src
+	register()
 
 /datum/asset/proc/register()
 	return
 
 /datum/asset/proc/send(client)
 	return
 
+
 //If you don't need anything complicated.
 /datum/asset/simple
+	_abstract = /datum/asset/simple
 	var/assets = list()
 	var/verify = FALSE
 
 /datum/asset/simple/register()
 	for(var/asset_name in assets)
 		register_asset(asset_name, assets[asset_name])
+
 /datum/asset/simple/send(client)
 	send_asset_list(client,assets,verify)
 
 
+// For registering or sending multiple others at once
+/datum/asset/group
+	_abstract = /datum/asset/group
+	var/list/children
+
+/datum/asset/group/register()
+	for(var/type in children)
+		get_asset_datum(type)
+
+/datum/asset/group/send(client/C)
+	for(var/type in children)
+		var/datum/asset/A = get_asset_datum(type)
+		A.send(C)
+
+
+// spritesheet implementation - coalesces various icons into a single .png file
+// and uses CSS to select icons out of that file - saves on transferring some
+// 1400-odd individual PNG files
+#define SPR_SIZE 1
+#define SPR_IDX 2
+#define SPRSZ_COUNT 1
+#define SPRSZ_ICON 2
+#define SPRSZ_STRIPPED 3
+
+/datum/asset/spritesheet
+	_abstract = /datum/asset/spritesheet
+	var/name
+	var/list/sizes = list()    // "32x32" -> list(10, icon/normal, icon/stripped)
+	var/list/sprites = list()  // "foo_bar" -> list("32x32", 5)
+	var/verify = FALSE
+
+/datum/asset/spritesheet/register()
+	if (!name)
+		CRASH("spritesheet [type] cannot register without a name")
+	ensure_stripped()
+
+	var/res_name = "spritesheet_[name].css"
+	var/fname = "data/spritesheets/[res_name]"
+	call("rust_g", "file_write")(generate_css(), fname)
+	register_asset(res_name, file(fname))
+
+	for(var/size_id in sizes)
+		var/size = sizes[size_id]
+		register_asset("[name]_[size_id].png", size[SPRSZ_STRIPPED])
+
+/datum/asset/spritesheet/send(client/C)
+	if (!name)
+		return
+	var/all = list("spritesheet_[name].css")
+	for(var/size_id in sizes)
+		all += "[name]_[size_id].png"
+	send_asset_list(C, all, verify)
+
+/datum/asset/spritesheet/proc/ensure_stripped(sizes_to_strip = sizes)
+	for(var/size_id in sizes_to_strip)
+		var/size = sizes[size_id]
+		if (size[SPRSZ_STRIPPED])
+			continue
+
+		// save flattened version
+		var/fname = "data/spritesheets/[name]_[size_id].png"
+		fcopy(size[SPRSZ_ICON], fname)
+		var/error = call("rust_g", "dmi_strip_metadata")(fname)
+		if(length(error))
+			stack_trace("Failed to strip [name]_[size_id].png: [error]")
+		size[SPRSZ_STRIPPED] = icon(fname)
+
+/datum/asset/spritesheet/proc/generate_css()
+	var/list/out = list()
+
+	for (var/size_id in sizes)
+		var/size = sizes[size_id]
+		var/icon/tiny = size[SPRSZ_ICON]
+		out += ".[name][size_id]{display:inline-block;width:[tiny.Width()]px;height:[tiny.Height()]px;background:url('[name]_[size_id].png') no-repeat;}"

```

Comment:
what's the browser support on all these? remember we have to support at least ie9 (potentially even ie8)

Reply:
also is there any advantage/possibility of doing this outside of byond?

Reply:
`display: inline-block` -- IE 8+
`width:` and `height:` -- IE 4+
`background: no-repeat` -- IE 4+
`background-position:` -- IE 4+

Possibility yes. Advantage unclear.